### PR TITLE
[Model] fix bugs in pygat and mxgat.

### PR DESCRIPTION
**Pull request by sufeidechabei**
cora: dgl pygat: 84.1% dgl mxgat: 82.9% original paper: 83.0 ± 0.7%
citeseer : dgl pygat : 70.10% dgl mxgat: 74% original paper: 72.5 ± 0.7%
pubmed: dgl pygat: 77.6% dgl mxgat: 75%  original paper: 79.0 ± 0.3%

### CODE REVIEW

File: examples/mxnet/gat/gat_batch.py [hunk 1]
```diff
@@ -168,39 +187,55 @@ def main(args):
                 args.attn_drop,
                 args.residual)
 
-    model.initialize(ctx=ctx)
+    model.initialize()
+
+    if cuda:
+        model.collect_params().reset_ctx(ctx)
```
Comment (ylfdq1118) [inline examples/mxnet/gat/gat_batch.py]: Why not just use the original way `model.initialize(ctx=ctx)`, which looks much cleaner

File: examples/mxnet/gat/gat_batch.py [hunk 2]
```diff
@@ -168,39 +187,55 @@ def main(args):
...
+            loss = mx.nd.softmax_cross_entropy(logits[mask].squeeze(), labels[mask].squeeze())
+            loss.backward()
+        trainer.step(mask.shape[0])
 
         if epoch >= 3:
             dur.append(time.time() - t0)
-            print("Epoch {:05d} | Loss {:.4f} | Time(s) {:.4f} | ETputs(KTEPS) {:.2f}".format(
-                epoch, loss.asnumpy()[0], np.mean(dur), n_edges / np.mean(dur) / 1000))
+        if epoch % 100 == 0:
+            val = logits[val_mask]
```
Comment (ylfdq1118) [inline examples/mxnet/gat/gat_batch.py]: There should be another run of forward for validation

File: examples/mxnet/gat/gat_batch.py [hunk 3]
```diff
@@ -168,39 +187,55 @@ def main(args):
...
             logits = model(features)
-            loss = mx.nd.softmax_cross_entropy(logits, labels)
-
-        #optimizer.zero_grad()
-        loss.backward()
-        trainer.step(features.shape[0])
-        loss.wait_to_read()
+            loss = mx.nd.softmax_cross_entropy(logits[mask].squeeze(), labels[mask].squeeze())
+            loss.backward()
+        trainer.step(mask.shape[0])
```
Comment (ylfdq1118) [inline examples/mxnet/gat/gat_batch.py]: This might be debatable @jermainewang . In gcn, batchsize in step call is set to 1. See here:
https://github.com/dmlc/dgl/blob/master/examples/mxnet/gcn/gcn.py#L180

So I am not sure whether we should rescale gradients by number of nodes here

File: examples/mxnet/gat/gat_batch.py [hunk 4]
```diff
@@ -212,9 +247,10 @@ def main(args):
             help="input feature dropout")
     parser.add_argument("--attn-drop", type=float, default=.6,
             help="attention dropout")
-    parser.add_argument("--lr", type=float, default=0.005,
+    parser.add_argument("--lr", type=float, default=0.01,
```
Comment (ylfdq1118) [inline examples/mxnet/gat/gat_batch.py]: lr is different from (larger than) pytorch version. Is it because the gradent rescale?

File: examples/pytorch/gat/gat.py [hunk 5]
```diff
@@ -190,30 +226,43 @@ def main(args):
         if epoch >= 3:
             dur.append(time.time() - t0)
 
+        if epoch % 100 == 0:
+            acc = evaluate(model, features, labels, val_mask)
+            model.train()
```
Comment (ylfdq1118) [inline examples/pytorch/gat/gat.py]: Put `model.train()` at the beginning of for loop

File: examples/mxnet/gat/gat_batch.py [hunk 6]
```diff
@@ -168,39 +187,55 @@ def main(args):
...
             dur.append(time.time() - t0)
-            print("Epoch {:05d} | Loss {:.4f} | Time(s) {:.4f} | ETputs(KTEPS) {:.2f}".format(
-                epoch, loss.asnumpy()[0], np.mean(dur), n_edges / np.mean(dur) / 1000))
+        if epoch % 100 == 0:
+            val = logits[val_mask]
+            val = val.asnumpy().squeeze()
+            val_labels = labels[val_mask].asnumpy().squeeze()
+            max_index = np.argmax(val, axis=1)
+            accuracy = np.sum(np.where(max_index == val_labels, 1, 0))/len(val_labels)
+            print("Validation accuracy: {:.2f}".format(accuracy))
```
Comment (ylfdq1118) [inline examples/mxnet/gat/gat_batch.py]: Move evaluation code into a separate function so that it can be shared by validation and testing

File: examples/mxnet/gat/gat_batch.py (hunk 3 above)
Comment (sufeidechabei) [reply examples/mxnet/gat/gat_batch.py]: I 'm sure we should rescale gradients.

**Comment by sufeidechabei** [comment]
Done @ylfdq1118

File: examples/mxnet/gat/gat_batch.py (hunk 3 above)
Comment (ylfdq1118) [reply examples/mxnet/gat/gat_batch.py]: Line 164-166 did take out all positions that mask != 0. The rescale should be correct. I just don't remember why we decided to use `trainer.step(batchsize=1)` in gcn. Maybe we also change GCN? @jermainewang

File: examples/pytorch/gat/gat.py [hunk 7]
```diff
@@ -122,6 +137,7 @@ def forward(self, features):
                 # prepare
                 self.g.ndata.update(self.prp[i](last))
                 # message passing
+
```
Comment (ylfdq1118) [inline examples/pytorch/gat/gat.py]: remove this empty line

File: examples/pytorch/gat/gat.py [hunk 8]
```diff
@@ -53,24 +58,34 @@ def forward(self, nodes):
...
+        nn.init.xavier_normal_(self.attn_r.weight.data, gain=1.414)
 
     def forward(self, feats):
         h = feats
-        if self.drop != 0.0:
-            h = F.dropout(h, self.drop)
+
+        if self.drop:
+            h = self.drop(h)
+
```
Comment (ylfdq1118) [inline examples/pytorch/gat/gat.py]: remove above two empty lines

File: examples/pytorch/gat/gat.py [hunk 9]
```diff
@@ -190,30 +227,42 @@ def main(args):
         if epoch >= 3:
             dur.append(time.time() - t0)
 
+        if epoch % 100 == 0:
```
Comment (ylfdq1118) [inline examples/pytorch/gat/gat.py]: move the validate after printing training loss

File: examples/pytorch/gat/gat.py [hunk 10]
```diff
@@ -151,9 +181,15 @@ def main(args):
         features = features.cuda()
         labels = labels.cuda()
         mask = mask.cuda()
+        val_mask = val_mask.cuda()
 
     # create GCN model
```
Comment (ylfdq1118) [inline examples/pytorch/gat/gat.py]: change this comment to something like "create dgl graph"

File: examples/pytorch/gat/gat.py [hunk 11]
```diff
@@ -151,9 +181,15 @@ def main(args):
...
     g = DGLGraph(data.graph)
+    # add self loop
+    g.add_edges(g.nodes(), g.nodes())
+
+
+
```
Comment (ylfdq1118) [inline examples/pytorch/gat/gat.py]: too many empty lines ...

Comment (sufeidechabei) [reply examples/pytorch/gat/gat.py]: OK @ylfdq1118

**Comment by jermainewang** [comment]
Yeh, I'm good. Go ahead merge it. Thanks everyone.
